extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -14,5 +14,71 @@
  * limitations under the License.
  */
 
+import {Messaging} from './messaging.js';
+import {listen} from '../../../src/event-helper';
+import {viewerForDoc} from '../../../src/viewer';
+import {user} from '../../../src/log';
+
+
 export class AmpViewerIntegration {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const @private {!Window} win */
+    this.win_ = win;
+  }
+
+  /**
+   * @return {!Promise}
+   */
+  init() {
+    return new Promise(resolve => {
+      const viewer = viewerForDoc(this.win_.document);
+      this.getHandshakePromise_(viewer)
+      .then(viewerOrigin => {
+        const messaging = new Messaging(this.win_.parent, viewerOrigin,
+            (type, payload, awaitResponse) => {
+              return viewer.receiveMessage(
+                type, /** @type {!JSONType} */ (payload), awaitResponse);
+            });
+        viewer.setMessageDeliverer((type, payload, awaitResponse) => {
+          return messaging.sendRequest(type, payload, awaitResponse);
+        }, viewerOrigin);
+        resolve();
+      });
+    });
+  }
+
+  /**
+   * @param {!../../../src/service/viewer-impl.Viewer} viewer
+   * @return {!Promise}
+   * @private
+   */
+  getHandshakePromise_(viewer) {
+    const win = this.win_;
+
+    return new Promise(resolve => {
+      const unconfirmedViewerOrigin = viewer.getParam('viewerorigin');
+      user().assert(unconfirmedViewerOrigin,
+              'Expected viewer origin must be specified!');
+
+      const listener = function(event) {
+        if (event.origin == unconfirmedViewerOrigin &&
+                event.data == 'amp-handshake-response' &&
```

Comment:
This should be +4 spaces from the `if` keyword (so, even with the `event.origin` above).